### PR TITLE
[ShapeOpt] Reorganizing optimizer_factory.py

### PR DESCRIPTION
**Pull request by adityaghantasala**
**Description**
I think the class `VertexMorphing` in the [optimizer_factory.py](https://github.com/KratosMultiphysics/Kratos/blob/master/applications/ShapeOptimizationApplication/python_scripts/optimizer_factory.py) is miss representing what is actually happening in the class. 

The changes also make the Optimizer class more "AnalysisStage" like.

**Changelog**
VertexMorphing Class is replaced by the class `Optimizer`. The validation of the parameters is also moved into this class. 
MainKratosStructuralOptimization.py is added as a guide for the main script.

Tests are also changed accordingly.

**IMPORTANT :  This will break the existing running cases**

### CODE REVIEW

File: applications/ShapeOptimizationApplication/python_scripts/optimizer_factory.py [hunk 1]
```diff
@@ -23,100 +23,28 @@
...
-        "reference"                           : "initial_value",
-        "reference_value"                     : 1.0,
-        "analyzer"                            : "external",
-        "response_settings"                   : {},
-        "project_gradient_on_surface_normals" : false
-    }""")
-    for itr in range(constraint_settings.size()):
-        constraint_settings[itr].ValidateAndAssignDefaults(default_settings)
-
+def Create( model, optimization_settings, external_analyzer=EmptyAnalyzer()):
```
Comment (armingeiser) [inline applications/ShapeOptimizationApplication/python_scripts/optimizer_factory.py:30]: we could keep the old `CreateOptimizer` function with a deprecation warning for backward compatibility reasons

**Comment by adityaghantasala** [comment]
@armingeiser Done ! as we discussed I put the CreateOptimizer function to maintain backward compatibility. Also renamed the main python script

**Review by armingeiser — APPROVED** [review]

File: applications/ShapeOptimizationApplication/python_scripts/optimizer_factory.py [hunk 2]
```diff
@@ -23,100 +23,32 @@
...
-        "response_settings"                   : {},
-        "project_gradient_on_surface_normals" : false
-    }""")
-    for itr in range(constraint_settings.size()):
-        constraint_settings[itr].ValidateAndAssignDefaults(default_settings)
+## Purely for backward compatibility, should be removed soon.
+def CreateOptimizer(optimization_settings,model,external_analyzer=EmptyAnalyzer()):
+    return Optimizer(model, optimization_settings, external_analyzer)
 
+def Create( model, optimization_settings, external_analyzer=EmptyAnalyzer()):
```
Comment (dbaumgaertner) [inline applications/ShapeOptimizationApplication/python_scripts/optimizer_factory.py:30]: Why changing the naming here? I like the changes in this PR, but breaking the existing API just for a minor rewording needs a good argument. I guess it's more "factory-like", right? I am fine with both, though. Just out of curiosity.

**Review by dbaumgaertner — APPROVED** [review]
Minor comment, otherwise I am approving it. Since I cannot test it right away. I rely on you and assume you made sure that the unit tests are running. Just double-check it, if not yet done already.

File: applications/ShapeOptimizationApplication/python_scripts/optimizer_factory.py (hunk 2 above)
Comment (adityaghantasala) [reply applications/ShapeOptimizationApplication/python_scripts/optimizer_factory.py:30]: The change is to be consistent with the other similar scripts across Kratos.

**Comment by adityaghantasala** [comment]
> Minor comment, otherwise I am approving it. Since I cannot test it right away. I rely on you and assume you made sure that the unit tests are running. Just double-check it, if not yet done already.

Yes I always run the test before pushing. Also since the app in in CD/CI the tests also are run and should pass so that a merge is allowed.

**Comment by adityaghantasala** [comment]
@dbaumgaertner @armingeiser Thank you guys.

**Comment by armingeiser** [comment]
> Also since the app in in CD/CI the tests also are run and should pass 

1/3 of the tests run in CI on the PR, the rest in nightly only.

**Pull request merged**